node/tessel-export.js

```diff
@@ -640,6 +719,29 @@ Tessel.Pin.prototype.analogWrite = function(val) {
   return this;
 };
 
+// Duty cycle should be a value between 0 and 1
+Tessel.Pin.prototype.pwmDutyCycle = function(dutyCycle, cb) {
+  // throw an error if this pin doesn't support PWM
+  if (!this.pwmSupported) {
+    throw new Error('PWM can only be used on G1 and G2 of either module port.');
+  }
+
+  if (typeof dutyCycle !== 'number' || dutyCycle > 1.0 || dutyCycle < 0) {
+    throw new Error('PWM duty cycle must be a number between 0 and 1');
+  }
+
+  // The frequency must be set prior to setting the duty cycle
+  if (Tessel.pwmPeriod === 0) {
+    throw new Error('PWM Frequency is not configured. You must call Tessel.pwmFrequency before setting duty cycle.');
+  }
+
+  var dutyCycleBuffer = new Buffer(2);
+  dutyCycleBuffer.writeUInt16BE(Math.floor(dutyCycle * Tessel.pwmPeriod));
+
+  this._port.sock.write(Buffer.concat([new Buffer([CMD.PWM_DUTY_CYCLE, this.pin]), dutyCycleBuffer]), cb);
+  return this;
+};
```

Comment:
```js
function measure(name, operation) {
  // First measure the execution time of just a function
  var hrcost = process.hrtime();
  (function() {}());
  var fcost = process.hrtime(hrcost);

  // Then measure execution time of the operation
  // wrapped in a function.
  var hrstart = process.hrtime();
  operation();
  var hrend = process.hrtime(hrstart);

  // Last, substract the cost of the function
  var s = hrend[0] - fcost[0];
  var ns = hrend[1] - fcost[1];

  console.info("%s: %ds %dms", name, s, ns / 1000000);
}

var PWM_DUTY_CYCLE = 27;
var dutyCycle = 1;

measure("a", function() {
  for (var i = 0; i < 1000; i++) {
    var dutyCycleBuffer = new Buffer(2);
    dutyCycleBuffer.writeUInt16BE(Math.floor(dutyCycle * 0xFF00));
    Buffer.concat([new Buffer([PWM_DUTY_CYCLE, 5]), dutyCycleBuffer])
  }
});

measure("b", function() {
  for (var i = 0; i < 1000; i++) {
    var data = Math.floor(dutyCycle * 0xFF00);
    new Buffer([PWM_DUTY_CYCLE, 5, data >> 8, data & 0xff]);
  }
});

```

Sample runs on a Mac: 

````
a: 0s 6.697157ms
b: 0s 0.658276ms

a: 0s 6.660633ms
b: 0s 0.642867ms

a: 0s 6.102375ms
b: 0s 0.664078ms

a: 0s 6.158544ms
b: 0s 0.684025ms

a: 0s 5.641681ms
b: 0s 0.756977ms

a: 0s 5.606084ms
b: 0s 0.696075ms
```

Sample runs on a Tessel 2: 

```
a: 0s 255.203632ms
b: 0s 82.86118ms

a: 0s 263.743691ms
b: 0s 84.681185ms

a: 0s 259.82363ms
b: 0s 83.621167ms

a: 0s 261.443648ms
b: 0s 86.461207ms

a: 0s 257.223584ms
b: 0s 83.461163ms

a: 0s 256.063563ms
b: 0s 82.481147ms

a: 0s 256.803567ms
b: 0s 82.241142ms

a: 0s 254.503527ms
b: 0s 82.721146ms

a: 0s 256.903555ms
b: 0s 85.721186ms
```


In every case, my optimization takes less than a tenth of a second (avg. ~83.8ms) to complete 1000 operations, where the current approach exceeds a quarter of a second (avg. ~257.97ms). 

Recommend replacing this: 

```js
var dutyCycleBuffer = new Buffer(2);
dutyCycleBuffer.writeUInt16BE(Math.floor(dutyCycle * Tessel.pwmPeriod));
this._port.sock.write(Buffer.concat([new Buffer([CMD.PWM_DUTY_CYCLE, this.pin]), dutyCycleBuffer]), cb);
```

With this: 

```js
var data = Math.floor(dutyCycle * Tessel.pwmPeriod); 
this._port.sock.write(new Buffer([CMD.PWM_DUTY_CYCLE,  this.pin, data >> 8, data & 0xff]), cb);
```